--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -327,8 +327,11 @@ install(FILES
 qt_add_resources(KF6WidgetsAddons "kcharselect-data"
     PREFIX "/kf6/kcharselect/"
     FILES kcharselect-data
+    OUTPUT_TARGETS resource_targets
 )
 
+install(TARGETS "${resource_targets}" EXPORT KF6WidgetsAddonsTargets OBJECTS DESTINATION "${CMAKE_INSTALL_LIBDIR}" COMPONENT Devel)
+
 if (BUILD_QCH)
     ecm_add_qch(
         KF6WidgetsAddons_QCH
