--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -21,11 +21,6 @@
 
 option(QT_INSTALL_XDG_DESKTOP_ENTRIES "Enable XDG desktop entries installation" OFF)
 
-if(TARGET Qt::Widgets)
-    add_subdirectory(uiplugin)
-    add_subdirectory(uitools)
-endif()
-
 add_subdirectory(global) # special case add as first directory
 if(QT_FEATURE_linguist)
     add_subdirectory(linguist)
--- a/src/linguist/CMakeLists.txt
+++ b/src/linguist/CMakeLists.txt
@@ -16,15 +16,7 @@
 add_subdirectory(lconvert)
 add_subdirectory(lcheck)
 add_subdirectory(lrelease)
-add_subdirectory(lrelease-pro)
 add_subdirectory(lupdate)
-add_subdirectory(lupdate-pro)
-add_subdirectory(ltext2id)
-if(QT_FEATURE_process AND QT_FEATURE_pushbutton AND QT_FEATURE_toolbutton
-    AND QT_FEATURE_png AND QT_FEATURE_mdiarea AND QT_FEATURE_syntaxhighlighter
-    AND TARGET Qt::Widgets AND TARGET Qt::Quick)
-    add_subdirectory(linguist)
-endif()
 
 # Create a fake module that would emulate the Qt5::LinguistTools CMake Config package
 qt_internal_add_module(Linguist
