diff --git a/src/corelib/CMakeLists.txt b/src/corelib/CMakeLists.txt
index d3f137847d..7eb92e64a6 100644
--- a/src/corelib/CMakeLists.txt
+++ b/src/corelib/CMakeLists.txt
@@ -814,11 +814,6 @@ qt_internal_extend_target(Core CONDITION QT_FEATURE_regularexpression
         WrapPCRE2::WrapPCRE2
 )
 
-qt_internal_extend_target(Core CONDITION QT_FEATURE_openssl_linked AND QT_FEATURE_opensslv30
-    LIBRARIES
-        WrapOpenSSL::WrapOpenSSL
-)
-
 qt_internal_extend_target(Core CONDITION QT_FEATURE_hijricalendar
     SOURCES
         time/qhijricalendar.cpp time/qhijricalendar_p.h
diff --git a/src/corelib/tools/qcryptographichash.cpp b/src/corelib/tools/qcryptographichash.cpp
index 6ec6de1307..6392d9297a 100644
--- a/src/corelib/tools/qcryptographichash.cpp
+++ b/src/corelib/tools/qcryptographichash.cpp
@@ -25,7 +25,7 @@
 #include "../../3rdparty/rfc6234/sha.h"
 
 #ifndef QT_CRYPTOGRAPHICHASH_ONLY_SHA1
-#if !QT_CONFIG(opensslv30) || !QT_CONFIG(openssl_linked)
+#if true
 // qdoc and qmake only need SHA-1
 #include "../../3rdparty/md5/md5.h"
 #include "../../3rdparty/md5/md5.cpp"
@@ -114,7 +114,7 @@ static inline int SHA384_512AddLength(SHA512Context *context, unsigned int lengt
 #endif
 #endif // QT_CRYPTOGRAPHICHASH_ONLY_SHA1
 
-#if !defined(QT_BOOTSTRAPPED) && QT_CONFIG(opensslv30) && QT_CONFIG(openssl_linked)
+#if false
 #define USING_OPENSSL30
 #include <openssl/evp.h>
 #include <openssl/provider.h>
